staging/src/k8s.io/apiserver/pkg/server/storage/resource_config.go

```diff
@@ -26,15 +31,38 @@ type APIResourceConfigSource interface {
 	AnyResourceForGroupEnabled(group string) bool
 }
 
+// GroupVersionRegistry provides access to registered group versions.
+type GroupVersionRegistry interface {
+	// IsGroupRegistered returns true if given group is registered.
+	IsGroupRegistered(group string) bool
+	// IsVersionRegistered returns true if given version is registered.
+	IsVersionRegistered(v schema.GroupVersion) bool
+	// PrioritizedVersionsAllGroups returns all registered group versions.
+	PrioritizedVersionsAllGroups() []schema.GroupVersion
+	// GroupVersionLifecycle returns the APILifecycle for the GroupVersion.
+	GroupVersionLifecycle(gv schema.GroupVersion) schema.APILifecycle
+	// ResourceLifecycle returns the APILifecycle for the GroupVersionResource.
+	ResourceLifecycle(gvr schema.GroupVersionResource) schema.APILifecycle
+}
+
 var _ APIResourceConfigSource = &ResourceConfig{}
 
 type ResourceConfig struct {
-	GroupVersionConfigs map[schema.GroupVersion]bool
-	ResourceConfigs     map[schema.GroupVersionResource]bool
+	GroupVersionConfigs  map[schema.GroupVersion]bool
+	ResourceConfigs      map[schema.GroupVersionResource]bool
+	compatibilityVersion *version.Version
+	GroupVersionRegistry
 }
 
-func NewResourceConfig() *ResourceConfig {
-	return &ResourceConfig{GroupVersionConfigs: map[schema.GroupVersion]bool{}, ResourceConfigs: map[schema.GroupVersionResource]bool{}}
+func NewResourceConfig(compatibilityVersion string, registry GroupVersionRegistry) *ResourceConfig {
+	// TODO: check featureGate.Enabled(CompatibilityVersion)
+	if compatibilityVersion == "" {
+		compatibilityVersion = baseversion.Get().GitVersion
+	}
```

Comment:
Could whatever code that passes in `compatibilityVersion` also handle falling back to binary version?

I'm imagining an approach where somewhere early on in configuration, the apiserver reads the --compatibility-version flag, and picks the "effective version" (which might be the binary version, or might be the compatibility version, depending on if --compatibility-version is set and if the feature flag for compatibility version is turned on), and then passes the effective version into the rest of the system, allowing us to avoid `if compatibilityVersion == "" { compatibilityVersion = binaryVersion }` throughout the rest of the code